### ISSUE TITLE
stable ssh keys

### ISSUE THREAD

**Issue by Sohalt** [astro/microvm.nix#52]
Since `/etc/ssh/` is on the tmpfs, the VM generates a new ssh key-pair every time is gets rebooted.
I tried to create a virtiofsd share for `/etc/ssh` to keep the keys on the host system, but that makes sshd fail (presumably because there is an issue with the symlinks for the config).
Any idea why this would fail? Symlinks should work across filesystems and the directory should be there from the moment the machine starts, no?

**Comment by astro** [comment]
I have my MicroVM's whole `/etc` on virtiofs on ZFS for that precise reason: stable ssh host keys. I guess the conflict is with NixOS' `/etc/static`?

If you think your solution is better practise we should add it to the handbook.

**Comment by yangm97** [comment]
I guess truly stable ssh keys would need to be generated beforehand and applied using some secret management tool i.e. https://github.com/Mic92/sops-nix

**Comment by Sohalt** [comment]
sops-nix uses the ssh host keys to decrypt all other secrets, so it's a chicken/egg problem

**Comment by yangm97** [comment]
> sops-nix uses the ssh host keys to decrypt all other secrets, so it's a chicken/egg problem

I was thinking about bootstrapping using sops-nix on the host to store the vm ssh key and then mount the file secret read-only.

**Comment by c0deaddict** [comment]
I use https://github.com/nix-community/impermanence/ to "solve" this problem. It can be used not only for SSH keys but also for other persistent state. I give each MicroVM a `/persist` shared via virtiofs, and then bind mount the relevant dirs/files:

 ```    services.openssh = {
      hostKeys = [
        {
          path = "/persist/etc/ssh/ssh_host_ed25519_key";
          type = "ed25519";
        }
        {
          path = "/persist/etc/ssh/ssh_host_rsa_key";
          type = "rsa";
          bits = 4096;
        }
      ];
    };

    fileSystems."/persist".neededForBoot = mkForce true;

    environment.persistence."/persist" = {
      directories = [
        "/var/lib/systemd/coredump"
        "/var/lib/nixos" # contains user/group id map
        "/var/log"
      ];

      files = [
        "/etc/machine-id"
        "/root/.bash_history"
      ];
    };```

**Comment by matthew-salerno** [comment]
I've managed to do this with [impermanence](https://github.com/nix-community/impermanence) and [agenix-rekey](https://github.com/oddlama/agenix-rekey). Disregarding boiler plate, it looks like this:

Host
```nix
config.age = {
  secrets.myMicroVMSsh = {
    owner = "root";
    mode = "400";
    group = "root";
    path = "/etc/vm-persist/myMicrovm/etc/ssh/ssh_host_ed25519_key";
    symlink = false;
    rekeyFile = ../secrets/myMicrovmSsh.age";
    generator.script = {pkgs, file, ...}: ''
      ${pkgs.openssh}/bin/ssh-keygen -qt ed25519 -N "" -C "root@${name}" -f ${lib.escapeShellArg (lib.removeSuffix ".age" file)}
      priv=$(${pkgs.coreutils}/bin/cat ${lib.escapeShellArg (lib.removeSuffix ".age" file)})
      ${pkgs.coreutils}/bin/shred -u ${lib.escapeShellArg (lib.removeSuffix ".age" file)}
      echo "$priv"
    '';
  };
};
```

Also Host (but defines the VM)
```
{ pkgs, impermanence, lib, agenix, agenix-rekey, ... }:
{
  microvm.vms.myMicrovm = {
    inherit pkgs;
    config = {
      imports = [
        agenix.nixosModules.default
        agenix-rekey.nixosModules.default
        impermanence.nixosModules.impermanence
      ];
      microvm.shares = [
          {
            source = "/etc/vm-persist/test-microvm";
            mountPoint = "/persist";
            tag = "persist";
            proto = "virtiofs";
          }
        ];
      age.rekey = {
        hostPubkey = ../secrets/myMicrovmSsh.pub;
        masterIdentities = [ ../../yubikey-ident.pub ];
      };
      services.openssh = {
        enable = true;
        hostKeys = [
          {
            path = "/etc/ssh/ssh_host_ed25519_key";
            type = "ed25519";
          }
        ];
      };
        
      fileSystems."/persist".neededForBoot = lib.mkForce true;
      environment.persistence."/persist" = {
        files = [
          "/etc/ssh/ssh_host_ed25519_key"
        ];
      };
    };
  };
}
```
I edited out the parts specific to my usecase (like configuring secrets for each vm in a map) and I haven't tested this code exactly, but the general idea works.

**Comment by RooSoft** [comment]
Steps I did:

- create a microVM
- ssh into it
- created a tarball containing all of `/etc/ssh`
- use [wormhole](https://github.com/magic-wormhole/magic-wormhole) to warp that tarball into the host
- on the host, extracted it in a `.ssh` folder near where I store .img files for the same microVM's volumes
- created a readonly share from that folder, straight into `/etc/ssh` on the microVM

From that point on, it's stable SSH.

**Comment by Bert-Proesmans** [comment]
I spent some time looking into the direction of authenticated seeding of secrets from host to vm. This is doable with VSOCK, systemd service unit magic, and socat. Adhoc hacking though, I haven't found a way to create nice and simple nixos module options to abstract for more secrets per vm.

After experimenting I'd say a solution with impermanence and a read/write mount is the least hassle. But not a perfect solution;
* mount has to be writable
* secrets touch filesystem
* immutable users and UID hassle

An authenticated runtime secret seeder that keeps and provides secrets in VM memory would be ideal for all security aspects. This becomes 'not simple' rather quick, and there is also the issue of handling secrets that need to be stored in a file for some programs (like sshd).
[Systemd has some direction](https://systemd.io/CREDENTIALS/#inheritance-from-container-managers-hypervisors-kernel-command-line-or-the-uefi-boot-environment) but that's all rather new to me. The suggestions aren't simple, requiring multiple moving parts (and their operational knowledge). I'll try to get a demo working, but don't wait up for me.

**Comment by astro** [comment]
Regarding systemd: there are plenty of new bells and whistles for virtual machines in v256 which I look forward to. Track it over here: https://github.com/NixOS/nixpkgs/pull/307068

**Comment by 39555** [comment]
@c0deaddict Sorry for necroreply. Do I understand correctly that all keys are also accesible from the host system in 'persist' directory?

**Comment by c0deaddict** [comment]
> @c0deaddict Sorry for necroreply. Do I understand correctly that all keys are also accesible from the host system in 'persist' directory?

Yes, that is correct. On my server each microvm gets `/persist/microvms/$hostname` (host) shared via virtiofs to `/persist` (vm)

**Comment by Ramblurr** [comment]
It should be possible to use systemd's smbious feature to pass secrets from the host to the guest, at least on qemu and on cloud-hypervisor.

Refs:
- https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4319
- https://github.com/systemd/systemd/blob/cf4deeaf1e822ade5c1fbbe2584b23a2d0988439/docs/CREDENTIALS.md#inheritance-from-container-managers-hypervisors-kernel-command-line-or-the-uefi-boot-environment

I tried to have a go at testing this for cloud-hypervisor, but unfortunately the `--platform` arg is already being passed, so it cannot be set again via `microvm.cloud-hypervisor.extraArgs`

https://github.com/astro/microvm.nix/blob/d3a9b7504d420a1ffd7c83c1bb8fe57deaf939d2/lib/runners/cloud-hypervisor.nix#L153

Assuming this works, I could easily imagine a small feature addition to microvm's nixos module that lets you define secrets to pass in.


``nix
  # bad way to do this for secrets, but still valid for non-secrets
  microvm.credentials.SOPS_AGE_KEY = "supersecret";
  # outputs => --platform oem_string=["io.systemd.credential:SOPS_AGE_KEY=supersecret"]

  # perhaps better way
  sops.secrets.guest_age_key = {
    owner = "microvm";
  };
  microvm.credentialFiles.SOPS_AGE_KEY = config.sops.secrets.guest_age_key.path;
```

Of course then the guest's sops systemd service would have to be configured to load the credential.

However the `--platform..` args will be visible with `ps` on the host, which seems less than desirable.

**Comment by SuperSandro2000** [comment]
You want to configure a share for /etc, see https://github.com/microvm-nix/microvm.nix/blob/main/doc/src/shares.md
